src/main/scala/shine/DPIA/showDPIA.scala

```diff
@@ -0,0 +1,86 @@
+  def tabulate : String => String =
+    _.linesIterator.map(l => if (l.isEmpty) l else TAB + l).mkString("\n")
+
+  def showTypeId : Kind.Identifier => String = _.name
+  def showKind : Kind[_, _] => String = _.name
+  def showUnary : Operators.Unary.Value => String = _.toString
+  def showBinary : Operators.Binary.Value => String = _.toString
+  def showData : Data => String = _.toString
+  def showNat : Nat => String = _.toString
+  def showPhraseType : PhraseType => String = _.toString
+  def showDataType : DataType => String = _.toString
+  def showAddressSpace : AddressSpace => String = _.toString
+  def showAccess : AccessType => String = _.toString
+  def showNatToNat : NatToNat => String = _.toString
+  def showNatToData : NatToData => String = _.toString
+  def showTypeDo[T](kind : Kind[T, _], arg : T) : String = kind match {
+    case PhraseKind => showPhraseType(arg.asInstanceOf[PhraseType])
+    case DataKind => showDataType(arg.asInstanceOf[DataType])
+    case NatKind => showNat(arg.asInstanceOf[Nat])
+    case AddressSpaceKind => showAddressSpace(arg.asInstanceOf[AddressSpace])
+    case AccessKind => showAccess(arg.asInstanceOf[AccessType])
+    case NatToNatKind => showNatToNat(arg.asInstanceOf[NatToNat])
+    case NatToDataKind => showNatToData(arg.asInstanceOf[NatToData])
+  }
+  def showPrim : Primitive[_] => String = {
+    case Comment(s) => s"/* ${s} */"
+    case Seq(c1, c2) => s"{\n${tabulate(showPhrase(c1))};\n${tabulate(showPhrase(c2))}\n}"
+    case New(dt, f) => s"new (exp[${showDataType(dt)}] x acc[${showDataType(dt)}]) in ${showPhrase(f)}"
+    case Assign(dt, lhs, rhs) => s"${showTypedPhrase(DataKind, dt, lhs)} = ${showPhrase(rhs)}"
+    case IdxAcc(n, dt, idx, array) => s"${showTypedPhrase(DataKind, ArrayType(n, dt), array)}[${showPhrase(idx)}]"
+    case Idx(n, dt, idx, array) => s"${showTypedPhrase(DataKind, ArrayType(n, dt), array)}[${showPhrase(idx)}]"
+    case NatAsIndex(n, e) => s"${showTypedPhrase(NatKind, n, e)}"
+    case Drop(n, m, dt, e) => s"drop ${showNat(n)} ${showTypedPhrase(DataKind, ArrayType(n + m, dt), e)}"
+    case DropAcc(n, m, dt, e) => s"drop ${showNat(n)} ${showTypedPhrase(DataKind, ArrayType(n + m, dt), e)}"
+    case Take(n, m, dt, e) => s"take ${showNat(n)} ${showTypedPhrase(DataKind, ArrayType(n + m, dt), e)}"
+    case TakeAcc(n, m, dt, e) => s"take ${showNat(n)} ${showTypedPhrase(DataKind, ArrayType(n + m, dt), e)}"
+    case f : For => s"for ${showNat(f.n)}\n${tabulate(showPhrase(f.loopBody))}\n"
+    case f : ForNat => s"forNat ${showNat(f.n)}\n${tabulate(showPhrase(f.loopBody))}\n"
+    case p => p.toString
+  }
+
+  def showKindedType[T](kind : Kind[T, _], t : T) : String = showTypes match {
+    case On => s"(${showTypeDo(kind, t)} : ${showKind(kind)})"
+    case Off => showTypeDo(kind, t)
+  }
+  def showTypedPhrase[T,U <: PhraseType](kind : Kind[T, _], t : T, phrase : Phrase[U]) : String = showTypes match {
+    case On => s"(${showPhrase(phrase)} : ${showTypeDo(kind, t)})"
```

Comment:
Do we not have `phrase.t == t`?

Reply:
True, I guess we do, since these terms were type checked. I'll simplify